### PR TITLE
PDF page to image conversion on upload

### PR DESCRIPTION
**Pull request by ajolipa**
### In this PR
An initial attempt at improving the performance of large PDFs by processing the individual pages into TIFF files on initial upload, and referencing those files in the generated manifest. Specifically, it makes the following changes:
- Adds a `pages_count` field to the `Resource` model for bookkeeping (NOTE: this is not currently relied on for anything, but it's not a lot of data and seems like it can't hurt to make this data easily accessible);
- Adds a `has_many_attached :content_converted_pages` attachment type to the Resource model, along with some related methods; this allows us to store multiple pages as related to a single parent PDF object.
- Adds a `convert_pdf` service that loops through the individual pages of the PDF and converts to TIFF.
  - This is where I had to fiddle a bunch to get the generated TIFF to render properly, and I still think there might be some loss of quality in how it's being first extracted from the PDF and then converted. Some of the trouble came from the way that PDFs handle transparency, which was leading to lingering data in the `alpha` channel that the JPEG compression couldn't handle. Adding `- alpha off` to the conversion seems to have fixed that error.
- Modifies the `create_manifest_job` to use the page-specific image links for the items, rather than relying on the `;{page_number}` suffix.

I should also mention that to get this working on my machine, I had to install `ImageMagick` and `Ghostscript` (I think that's what it was called?), which are used for parts of the PDF processing. Possibly an update to the Docker script to make sure those are included on the server would also be necessary, if they're not there by default.

Disclosure: Some of this was done with the help of the copilot chat thingy in VSCode, for help with proper Ruby syntax and some initial suggestions about how to incorporate the multi-page PDF concept into the content model, and handling some of the new URL methods. I've been through all of the code (I had to, since a lot of it did not work...) and I've gotten it to the point of successfully processing a PDF in my local environment and showing it in the IIIF viewer with the pages pointing at the newly generated TIFFs. As mentioned I have some lingering concerns about the quality of the individual page images, and also we would need to very thoroughly test corner cases and try it out with a wide range of PDFs. We should also see if we can actually quantify any performance improvements over the old method. But it's a start. ^_^;

### CODE REVIEW

File: app/services/images/convert_pdf.rb [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        output_filename = "#{base_filename}_page_#{page_number + 1}.#{output_format}"
+        output_path = File.join(File.dirname(file.path), output_filename)
+
+        # Use ImageMagick to extract the specific PDF page
+        convert = MiniMagick.convert
+        convert << "#{file.path}[#{page_number}]"  # Specify page index (0-indexed)
+        convert << '-quality'
+        convert << '90'  # Good quality for intermediate conversion
+        convert << '-density'
+        convert << '300'  # High DPI for better quality
```
Comment (blms) [inline app/services/images/convert_pdf.rb]: Caught by Claude—apparently this `-density 300` flag has to come _before_ the file path. No idea why it's like that, and [apparently others have run into this](https://github.com/minimagick/minimagick/issues/442).

File: app/jobs/convert_image_job.rb [hunk 2]
```diff
@@ -5,35 +5,116 @@ class ConvertImageJob < ApplicationJob
...
+            # Continue with next page rather than failing entire job
+          rescue Exceptions::PDFPageConversionError => e
+            Rails.logger.error "Failed to convert page #{page_number} to TIFF for resource #{resource.id}: #{e.message}"
+            # Continue with next page rather than failing entire job
+          end
+        end
+
+        # Store page count on resource for tracking
+        resource.update(pages_count: page_count)
+
```
Comment (blms) [inline app/jobs/convert_image_job.rb]: Should we also regenerate the manifest here? Like `CreateManifestJob.perform_later(resource.id)`. That way the manifest gets all of the converted image URLs.

File: app/controllers/public/resources_controller.rb [hunk 3]
```diff
@@ -25,7 +25,13 @@ def iiif
     page_number = params[:page] || 1
 
     redirect_resource do |resource|
-      resource.image? ? resource.content_converted_iiif_url(page_number) : resource.content_iiif_url(page_number)
+      if resource.converted_pages?
```
Comment (blms) [inline app/controllers/public/resources_controller.rb]: Do the other endpoints in this file need the `resource.converted_pages?` check and fork to use the new URLs you defined in `resource.rb`? 

e.g. `image_api` can use `content_converted_pages_image_api_url`, `info` can use `content_converted_pages_info_url`, same sort of thing for `preview` and `thumbnail`

File: app/services/images/convert_pdf.rb (hunk 1 above)
Comment (ajolipa) [reply app/services/images/convert_pdf.rb]: Wow this was genius and I would never have caught it; switching the order definitely fixed the resolution!

File: app/jobs/convert_image_job.rb (hunk 2 above)
Comment (ajolipa) [reply app/jobs/convert_image_job.rb]: Good call; done.

File: app/controllers/public/resources_controller.rb (hunk 3 above)
Comment (ajolipa) [reply app/controllers/public/resources_controller.rb]: Good point; I've added all the ones that seem relevant but I admit to not having a thorough sense of all the contexts in which these endpoints are used so I haven't tested fully. Thumbnails and previews seem to work as expected at least.

File: app/controllers/public/resources_controller.rb [hunk 4]
```diff
@@ -25,30 +25,53 @@ def iiif
...
+      end
     end
   end
 
   def info
     page_number = params[:page] || 1
 
     redirect_resource do |resource|
-      resource.image? ? resource.content_converted_info_url(page_number) : resource.content_info_url(page_number)
+      if resource.converted_pages
```
Comment (blms) [inline app/controllers/public/resources_controller.rb]:
```suggestion
      if resource.converted_pages?
```
typo?

File: app/controllers/public/resources_controller.rb [hunk 5]
```diff
@@ -25,30 +25,53 @@ def iiif
...
+        resource.content_converted_pages_info_url(page_number)
```
Comment (blms) [inline app/controllers/public/resources_controller.rb]: Should we do `page_number.to_i` everywhere the new helpers are called, since it does a numeric comparison after passing the arg to `content_converted_pages_base_url`?

File: app/controllers/public/resources_controller.rb (hunk 3 above)
Comment (blms) [reply app/controllers/public/resources_controller.rb]: - `info`: redirects to info.json, the IIIF Image Information doc (dimensions, tile sizes, formats). Viewers fetch this once before requesting any tiles
- `image_api`: the full IIIF Image API endpoint (with params like size, rotation, etc). This is the most-fetched endpoint by IIIF viewers with deep zoom capabilities
- `iiif`: convenience redirect to an image or page as a JPEG (at full max res `/full/max/0/default.jpg`)
- `preview` and `thumbnail` are pretty self explanatory, they are fixed size JPEG derivatives

File: app/controllers/public/resources_controller.rb (hunk 5 above)
Comment (ajolipa) [reply app/controllers/public/resources_controller.rb]: Good catch; seems safer to me to just do the conversion inside `content_converted_pages_base_url` rather than having to double check the type of `page_number` everywhere any of these methods are called, so I added that in, but if you think it would be better to convert each argument before passing it we can do that instead (or in addition).

**Review by blms — COMMENTED** [review]
Looking really good!

Sorry to keep going back and forth on this one, I have two remaining questions: 
1. If a page gets skipped during conversion due to an error of some kind, will the page numbering get messed up? I'm wondering if we ought to determine page number by filename instead of array position. Then a failed page could become a gap instead of shifting all later pages to earlier page numbers.
2. Are there performance implications of processing each page in a separate MiniMagick call as opposed to processing them all in one? It seems like it could be slow to do all those separate calls, though OTOH loading a large PDF all at once could also cause memory problems… maybe the ideal would be some kind of chunked processing where it does a predefined N pages at a time and then stitches them all together? 

   It's probably fine as is for now, but let's keep an eye on performance for large PDFs and look at that as a potential optimization later.

**Review by blms — APPROVED** [review]
Looks good!
